interfaces/polkit/backend.go

```diff
@@ -55,64 +61,90 @@ func (b *Backend) Name() interfaces.SecuritySystem {
 	spec, err := repo.SnapSpecification(b.Name(), appSet, opts)
 	if err != nil {
 		return fmt.Errorf("cannot obtain polkit specification for snap %q: %s", snapName, err)
 	}
 
 	// Get the files that this snap should have
-	glob := polkitPolicyName(snapName, "*")
-	content := deriveContent(spec.(*Specification), appSet)
-	dir := dirs.SnapPolkitPolicyDir
+	policiesContent, rulesContent := deriveContent(spec.(*Specification), appSet)
 
 	// If we do not have any content to write, there is no point
 	// ensuring the directory exists.
-	if content != nil {
-		if err := os.MkdirAll(dir, 0755); err != nil {
-			return fmt.Errorf("cannot create directory for polkit policy files %q: %s", dir, err)
+	if policiesContent != nil {
+		if err := os.MkdirAll(dirs.SnapPolkitPolicyDir, 0755); err != nil {
+			return fmt.Errorf("cannot create directory for polkit policy files %q: %s", dirs.SnapPolkitPolicyDir, err)
 		}
 	}
-	_, _, err = osutil.EnsureDirState(dir, glob, content)
+	glob := polkitPolicyName(snapName, "*")
+	_, _, err = osutil.EnsureDirState(dirs.SnapPolkitPolicyDir, glob, policiesContent)
 	if err != nil {
 		return fmt.Errorf("cannot synchronize polkit policy files for snap %q: %s", snapName, err)
 	}
+
+	if rulesContent != nil {
+		if err := os.MkdirAll(dirs.SnapPolkitRuleDir, 0755); err != nil {
+			return fmt.Errorf("cannot create directory for polkit rule files %q: %s", dirs.SnapPolkitRuleDir, err)
+		}
+	}
+	glob = polkitRuleName(snapName, "*")
+	_, _, err = osutil.EnsureDirState(dirs.SnapPolkitRuleDir, glob, rulesContent)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize polkit rule files for snap %q: %s", snapName, err)
+	}
+
 	return nil
 }
 
-// Remove removes polkit policy files of a given snap.
+// Remove removes polkit policy and rule files of a given snap.
 //
 // This method should be called after removing a snap.
 func (b *Backend) Remove(snapName string) error {
+	// Removal must be best-effort to avoid leaving dangling files on early errors.
 	glob := polkitPolicyName(snapName, "*")
-	_, _, err := osutil.EnsureDirState(dirs.SnapPolkitPolicyDir, glob, nil)
-	if err != nil {
-		return fmt.Errorf("cannot synchronize polkit files for snap %q: %s", snapName, err)
+	_, _, policyErr := osutil.EnsureDirState(dirs.SnapPolkitPolicyDir, glob, nil)
+	glob = polkitRuleName(snapName, "*")
+	_, _, ruleErr := osutil.EnsureDirState(dirs.SnapPolkitRuleDir, glob, nil)
+	if policyErr != nil || ruleErr != nil {
+		return fmt.Errorf("cannot synchronize polkit files for snap %q: %s", snapName, strutil.JoinErrors(policyErr, ruleErr))
 	}
 	return nil
 }
 
 // deriveContent combines security snippets collected from all the interfaces
 // affecting a given snap into a content map applicable to EnsureDirState.
-func deriveContent(spec *Specification, appSet *interfaces.SnapAppSet) map[string]osutil.FileState {
+func deriveContent(spec *Specification, appSet *interfaces.SnapAppSet) (policiesContent, rulesContent map[string]osutil.FileState) {
```

Comment:
my preference would be to have two helper here, once for policies and one for rules, I don't see a big win to have a single helper, also the diff would be smaller

Reply:
Thanks, Updated!